File: sim/event_loop.py
from typing import Dict, Any, List


class Listener:
    pass


class EventLoop:
    def __init__(self) -> None:
        self.listeners: Dict[str, List[Listener]] = dict()

    def add(self, listener: Listener, events: List[str]) -> None:
        if isinstance(events, str):
            events = [events]
        for event in events:
            if event not in self.listeners:
                self.listeners[event] = []
            self.listeners[event].append(listener)

    def remove(self, listener: Listener) -> None:
        for event in self.listeners:
            if listener in self.listeners[event]:
                self.listeners[event].remove(listener)

    def emit(self, event: str, *args, **kwargs) -> None:
        listeners = self.listeners[event] if event in self.listeners else []
        for listener in listeners:
            func = getattr(listener, event)
            func(*args, **kwargs)

File: src/util/EventLoop.ts
type Listener<Data> = (data: Data) => void

type ListenerMap<
    Event extends string,
    EventMapping extends Record<Event, any>
> = {
        [eventName in Event]?: Set<Listener<EventMapping[eventName]>>
    }

export class EventLoop<
    Event extends string,
    EventMapping extends Record<Event, any>
> {
    private listeners: ListenerMap<Event, EventMapping> = {}

    on<SpecificEvent extends Event>(
        event: SpecificEvent,
        listener: Listener<EventMapping[SpecificEvent]>
    ) {
        if (this.listeners[event]) {
            this.listeners[event].add(listener)
        } else {
            this.listeners[event] = new Set([listener])
        }
    }

    removeListener<SpecificEvent extends Event>(
        event: SpecificEvent,
        listener: Listener<EventMapping[SpecificEvent]>
    ) {
        if (this.listeners[event]) {
            this.listeners[event].delete(listener)
        }
    }

    emit<SpecificEvent extends Event>(
        event: SpecificEvent,
        data: EventMapping[SpecificEvent]
    ) {
        if (this.listeners[event]) {
            for (const listener of this.listeners[event]) {
                listener(data)
            }
        }
    }
}
